src/events/dns.js

```diff
@@ -0,0 +1,239 @@
+                } else {
+                    callbackArg = { records: arg };
+                }
+                endDurationAndAddCallbackMetadata(dnsEvent, startTime, err, { ...callbackArg });
+                resolve();
+                if (callback) callback(err, arg);
+            };
+        });
+        clientRequest = original.apply(this, [hostname, rrtype, patchedCallback]);
+        tracer.addEvent(dnsEvent, responsePromise);
+    } catch (err) {
+        tracer.addException(err);
+    }
+    if (!clientRequest) {
+        clientRequest = original.apply(this, buildParams(arg1, arg2, arg3));
+    }
+    return clientRequest;
+};
+
+const wrapDnsLookupServiceFunction = original => (address, port, callback) => {
+    let patchedCallback;
+    let clientRequest;
+    if (!callback) return original.apply(this, buildParams(address, port, callback));
+    try {
+        const { dnsEvent, startTime } = initialDnsEvent(original.name);
+        eventInterface.addToMetadata(dnsEvent, { address, port });
+        const responsePromise = new Promise((resolve) => {
+            patchedCallback = (err, hostname, service) => {
+                endDurationAndAddCallbackMetadata(dnsEvent, startTime, err, { hostname, service });
+                resolve();
+                if (callback) callback(err, hostname, service);
+            };
+        });
+        clientRequest = original.apply(this, [address, port, patchedCallback]);
+        tracer.addEvent(dnsEvent, responsePromise);
+    } catch (err) {
+        tracer.addException(err);
+    }
+    if (!clientRequest) {
+        clientRequest = original.apply(this, buildParams(address, port, callback));
+    }
+    return clientRequest;
+};
+
+const wrapDnsReverseFunction = original => (ip, callback) => {
+    let patchedCallback;
+    let clientRequest;
+    try {
+        const { dnsEvent, startTime } = initialDnsEvent(original.name);
+        eventInterface.addToMetadata(dnsEvent, { ip });
+        const responsePromise = new Promise((resolve) => {
+            patchedCallback = (err, hostnames) => {
+                endDurationAndAddCallbackMetadata(dnsEvent, startTime, err, { hostnames });
+                resolve();
+                if (callback) callback(err, hostnames);
+            };
+        });
+        clientRequest = original.apply(this, [ip, patchedCallback]);
+        tracer.addEvent(dnsEvent, responsePromise);
+    } catch (err) {
+        tracer.addException(err);
+    }
+    if (!clientRequest) {
+        clientRequest = original.apply(this, [ip, callback]);
+    }
+    return clientRequest;
+};
+
+const wrapDnsLookupFunction = original => (arg1, arg2, arg3) => {
+    let patchedCallback;
+    let clientRequest;
+    const { hostname, options, callback } = getArguments(arg1, arg2, arg3);
+    if (isBlacklistURL(hostname)) {
+        utils.debugLog(`filtered blacklist hostname ${hostname}`);
+        return original.apply(this, buildParams(arg1, arg2, arg3));
```

Comment:
just FYI, you can use the `arguments` keyword here and in similar places (but the rest of the instrumentation doesn't so take the suggestion I put on the actual buildParams method instead, just wanted you to be acquainted with it)